### PR TITLE
Feature/variables data source

### PR DESCRIPTION
**Pull request by kilwa0**
## Description

Hi all, this change add the variables data source. It outputs into a list some useful values from the configured variables on a given worskpace

It outputs three values:
  1. `variables` - the whole set
  1. `terraform` - only the terraform categorized variables
  1. `environment` - the environment variables into the worskpace

As you can tell English is not my primary language so please double check documentation for misspelled words :)
Please let me know if you need anything from me

Regards

## Testing plan

Acceptance test creates:

```
resource "tfe_organization" "foobar" {
  name  = "org-%d"
  email = "admin@company.com"
}

resource "tfe_workspace" "foobar" {
  name         = "workspace-foo-%d"
  organization = tfe_organization.foobar.id
}

resource "tfe_variable" "terrbar" {
	key          = "foo"
	value        = "bar"
	category     = "terraform"
	workspace_id = tfe_workspace.foobar.id
}

resource "tfe_variable" "envbar" {
	key          = "foo"
	value        = "bar"
	category     = "env"
	workspace_id = tfe_workspace.foobar.id
}

data "tfe_variables" "foobar" {
	workspace_id = tfe_workspace.foobar.id
	depends_on = [
    tfe_variable.terrbar,
		tfe_variable.envbar
  ]
}

output "variables" {
	value = data.tfe_variables.foobar.variables[0]["name"]
}

output "environment" {
	value = data.tfe_variables.foobar.environment[0]["name"]
}

output "terraform" {
	value = data.tfe_variables.foobar.terraform[0]["name"]
}
```

End checks the id and the outputs of the data source

## Output from acceptance tests

```
TESTARGS="-run TestAccTFEVariablesDataSource_basic" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run TestAccTFEVariablesDataSource_basic -timeout 15m
?   	github.com/hashicorp/terraform-provider-tfe	[no test files]
=== RUN   TestAccTFEVariablesDataSource_basic
--- PASS: TestAccTFEVariablesDataSource_basic (19.16s)
PASS
ok  	github.com/hashicorp/terraform-provider-tfe/tfe	19.168s
?   	github.com/hashicorp/terraform-provider-tfe/version	[no test files]
```

### CODE REVIEW

File: tfe/data_source_variables.go
Comment (brandonc) [inline tfe/data_source_variables.go:156]: Since these two conditions are identical, you can use the condition to figure out which slice to append result to:

I would also add "sensitive" to the schema and set that:

```suggestion
                      result := make(map[string]interface{})
                      result["id"] = variable.ID
                      result["name"] = variable.Key
                      result["category"] = variable.Category
                      result["hcl"] = variable.HCL
                      result["sensitive"] = variable.Sensitive
                      result["value"] = variable.Value
                      if variable.Category == "terraform" {
                              terraformVars = append(terraformVars, result)
                      } else if variable.Category == "env" {
                              envVars = append(envVars, result)
                      }
```

File: tfe/data_source_variables.go
Comment (brandonc) [inline tfe/data_source_variables.go:176]: ☝️ probably get rid of this

File: tfe/data_source_variables.go
Comment (brandonc) [inline tfe/data_source_variables.go:159]: I think it would be better to not interleave pages of terraform and env variables in this collection. At the end, you can just append totalEnvVariables to totalTerraformVariables (see below) so that all the terraform variables come first and all the env variables come after. It also simplifies the code and reduces copying.
```suggestion
```

File: tfe/data_source_variables.go
Comment (brandonc) [inline tfe/data_source_variables.go:173]:
```suggestion
	d.Set("variables", append(totalTerraformVariables, totalEnvVariables...))
```

File: tfe/data_source_variables.go
Comment (brandonc) [inline tfe/data_source_variables.go:112]: use camelCase

```suggestion
	workspaceID := d.Get("workspace_id").(string)
```

File: tfe/data_source_variables.go
Comment (brandonc) [inline tfe/data_source_variables.go:20]: Since this schema is defined 3x, it might be nice to just extract it once and reuse it-- I also removed the redundant `&schema.Schema` declaration

```
varSchema := map[string]*schema.Schema{
    "id": {
	    Type:     schema.TypeString,
	    Computed: true,
    },
    "name": {
	    Type:     schema.TypeString,
	    Computed: true,
    },
    "value": {
	    Type:      schema.TypeString,
	    Computed:  true,
	    Sensitive: true,
    },
    "category": {
	    Type:     schema.TypeString,
	    Computed: true,
    },
    "hcl": {
	    Type:     schema.TypeBool,
	    Computed: true,
    },
    "sensitive": {
	    Type:     schema.TypeBool,
	    Computed: true,
    },
}
```

File: website/docs/d/variables.html.markdown
Comment (brandonc) [inline website/docs/d/variables.html.markdown:22]:
```suggestion
  workspace_id = data.tfe_workspace.test.id
```

File: website/docs/d/variables.html.markdown
Comment (brandonc) [inline website/docs/d/variables.html.markdown:11]:
```suggestion
This data source is used to retrieve all variables defined in a specified workspace
```

File: website/docs/d/variables.html.markdown
Comment (brandonc) [inline website/docs/d/variables.html.markdown:30]:
```suggestion
* `workspace_id` - (Required) ID of the workspace.
```

File: website/docs/d/variables.html.markdown
Comment (brandonc) [inline website/docs/d/variables.html.markdown:34]:
```suggestion
* `variables` - List containing all terraform and environment variables configured on the workspace
```

File: website/docs/d/variables.html.markdown
Comment (brandonc) [inline website/docs/d/variables.html.markdown:35]:
```suggestion
* `terraform` - List containing terraform variables configured on the workspace
```

File: website/docs/d/variables.html.markdown
Comment (brandonc) [inline website/docs/d/variables.html.markdown:36]:
```suggestion
* `environment` - List containing environment variables configured on the workspace
```

File: website/docs/d/variables.html.markdown
Comment (brandonc) [inline website/docs/d/variables.html.markdown:44]:
```suggestion
* `sensitive` - If the variable is marked as sensitive or not
* `hcl` - If the variable is marked as HCL or not
```

File: website/docs/d/variables.html.markdown
Comment (brandonc) [inline website/docs/d/variables.html.markdown:42]:
```suggestion
* `value` -  The variable value. If the variable is sensitive this value will be empty.
```

**Review by brandonc — COMMENTED** [review]
Thanks for implementing this! It looks very promising. Here's my initial code review. Tomorrow I'll take some time to smoke test and ensure everything works as I imagine it did.

File: tfe/data_source_variables.go
Comment (kilwa0) [reply tfe/data_source_variables.go:112]: refactored in the incoming commit

File: tfe/data_source_variables.go
Comment (kilwa0) [reply tfe/data_source_variables.go:125]: ty

File: tfe/data_source_variables.go
Comment (kilwa0) [reply tfe/data_source_variables.go:136]: of course you are right, thanks

File: tfe/data_source_variables.go
Comment (kilwa0) [reply tfe/data_source_variables.go:156]: big commit incoming solving also this

File: tfe/data_source_variables.go
Comment (kilwa0) [reply tfe/data_source_variables.go:159]: :thumbsup:

File: tfe/data_source_variables.go
Comment (kilwa0) [reply tfe/data_source_variables.go:173]: :man_facepalming:

File: tfe/data_source_variables.go
Comment (kilwa0) [reply tfe/data_source_variables.go:176]: yeah forgot that xD

**Comment by brandonc** [comment]
@kilwa0 Thanks for making my recommended changes! I didn't have time to finalize my testing yet but this is still on my radar.

File: tfe/data_source_variables.go
Comment (brandonc) [inline tfe/data_source_variables.go:43]: As I was using the resource, I found myself expecting this to mirror the category and be named "env". Could you rename this schema field (as well as in the docs?)

File: tfe/data_source_variables.go
Comment (brandonc) [inline tfe/data_source_variables.go:42]: The data source works well! I think in the future we might want to add terraform_keys and env_keys as fields to enable other uses, but if this suits you for now it doesn't need to be added right away.

**Review by brandonc — CHANGES_REQUESTED** [review]
One final tweak! Thanks again

File: tfe/data_source_variables.go
Comment (kilwa0) [reply tfe/data_source_variables.go:42]: @brandonc 
sorry the dealy the same day I did my last commit went to holidays 😄. It's great you have time to test it thanks for your help

File: tfe/data_source_variables.go
Comment (kilwa0) [reply tfe/data_source_variables.go]: Did the changes and the tests output went well but please review them because I'm working from my moms setup and i'm not comfortable on it, sorry about that and thanks, @brandonc

Comment (kilwa0) [reply tfe/data_source_variables.go]: Todayt I was able to complete all my tests it looks fine to me
Regards

**Comment by brandonc** [comment]
@kilwa0 I'm ready to merge this! Can you give me write access on the forked repo so I can get this running on our CI to unblock the merge?

**Comment by kilwa0** [comment]
@brandonc I've sent you an invitation as collaborator.Iif that is not what you need tell me because I have no idea how to manage access on GH 😄

**Comment by brandonc** [comment]
I think I am actually supposed to push this branch to our repo instead. Sorry for the confusion!

**Comment by kilwa0** [comment]
> I think I am actually supposed to push this branch to our repo instead. Sorry for the confusion!

Allow edit from maintainers checkbox is checked

**Review by brandonc — APPROVED** [review]

**Comment by brandonc** [comment]
@kilwa0 Congrats and thanks for the excellent contribution. You're always welcome to submit more changes that simplify your workflow.

**Comment by kilwa0** [comment]
@brandonc My pleasure.
Thank you for your patience and guidance. See you soon :+1:

**Pull request merged**
